migrate_cpp/cpp_refactoring/rewriter.cpp

```diff
@@ -0,0 +1,286 @@
+// not be semantically correct: If the C++ code specifies the type `long`, and
+// on the platform for which the migration is occurring `long` has 64-bits, we
+// may not want to use `i64` as the replacement: The C++ code may be intended to
+// operate in environments where `long` is only 32-bits wide. We need to develop
+// a strategy for determining builtin-type replacements that addresses these
+// issues.
+auto RewriteBuilder::VisitBuiltinTypeLoc(clang::BuiltinTypeLoc type_loc)
+    -> bool {
+  std::string_view content;
+  switch (type_loc.getTypePtr()->getKind()) {
+    case clang::BuiltinType::Bool:
+      content = "bool";
+      break;
+    case clang::BuiltinType::Char_U:
+      content = "char";
+      break;
+    case clang::BuiltinType::UChar:
+      content = "u8";
+      break;
+    case clang::BuiltinType::UShort:
+      content = "u16";
+      break;
+    case clang::BuiltinType::UInt:
+      content = "u32";
+      break;
+    case clang::BuiltinType::ULong:
+      content = "u64";
+      break;
+    case clang::BuiltinType::ULongLong:
+      content = "u64";
+      break;
+    case clang::BuiltinType::UInt128:
+      content = "u128";
+      break;
+    case clang::BuiltinType::Char_S:
+      content = "char";
+      break;
+    case clang::BuiltinType::SChar:
+      content = "i8";
+      break;
+    case clang::BuiltinType::Short:
+      content = "i16";
+      break;
+    case clang::BuiltinType::Int:
+      content = "i32";
+      break;
+    case clang::BuiltinType::Long:
+      content = "i64";
+      break;
+    case clang::BuiltinType::LongLong:
+      content = "i64";
+      break;
+    case clang::BuiltinType::Int128:
+      content = "i128";
+      break;
+    case clang::BuiltinType::Float:
+      content = "f32";
+      break;
+    case clang::BuiltinType::Double:
+      content = "f64";
+      break;
+    default:
+      // In this case we do not know what the output should be so we do not
+      // write any.
+      return true;
+  }
+  Write(type_loc, OutputSegment::Text(content));
+  return true;
+}
+
+auto RewriteBuilder::VisitCXXBoolLiteralExpr(clang::CXXBoolLiteralExpr* expr)
+    -> bool {
+  Write(expr, OutputSegment::Text(expr->getValue() ? "true" : "false"));
+  return true;
+}
+
+auto RewriteBuilder::VisitDeclRefExpr(clang::DeclRefExpr* expr) -> bool {
+  Write(expr, OutputSegment::Text(TextForTokenAt(expr->getBeginLoc())));
+  return true;
+}
+
+auto RewriteBuilder::VisitDeclStmt(clang::DeclStmt* stmt) -> bool {
+  std::vector<OutputSegment> segments;
+  for (clang::Decl* decl : stmt->decls()) {
+    segments.push_back(OutputSegment::Node(decl));
+    segments.push_back(OutputSegment::Text(";\n"));
+  }
+  Write(stmt, std::move(segments));
+  return true;
+}
+
+auto RewriteBuilder::VisitIntegerLiteral(clang::IntegerLiteral* expr) -> bool {
+  // TODO: Replace suffixes.
+  std::string text(TextForTokenAt(expr->getBeginLoc()));
+  for (char& c : text) {
+    if (c == '\'') {
```

Comment:
```suggestion
    // 1'000 -> 1_000
    if (c == '\'') {
```

This made me pause to think about what was going on.

Reply:
Done.